packages/eslint-plugin/tests/rules/prefer-function-type.test.ts

```diff
@@ -129,5 +130,65 @@ interface Foo<T> {
       output: `
 type Foo<T> = (bar: T) => string;`,
     },
+    {
+      code: 'interface Foo { (): () => this; }',
+      errors: [
+        {
+          messageId: 'functionTypeOverCallableType',
+          tyoe: AST_NODE_TYPES.TSCallSignatureDeclaration,
+        },
+      ],
+      output: `type Foo = () => () => Foo;`,
+    },
+    {
+      code: 'interface Foo { (): () => string; }',
+      errors: [
+        {
+          messageId: 'functionTypeOverCallableType',
+          tyoe: AST_NODE_TYPES.TSCallSignatureDeclaration,
+        },
+      ],
+      output: `type Foo = () => () => string;`,
+    },
+    {
+      code: 'interface Interface {(): this}',
+      errors: [
+        {
+          messageId: 'functionTypeOverCallableType',
+          tyoe: AST_NODE_TYPES.TSCallSignatureDeclaration,
+        },
+      ],
+      output: `type Interface = () => Interface`,
+    },
+    {
+      code: 'interface Interface {(): Object<this>}',
```

Comment:
@bradzacher  this ( generic types ) will fail as the latest commit is still WIP.